Utils/Nginx/config.h

```diff
@@ -0,0 +1,207 @@
+CURRENT_FORWARD_DECLARE_STRUCT(LocationDirective);
+
+using directives_t = Variant<SimpleDirective, BlockDirective, ServerDirective, LocationDirective>;
+
+CURRENT_STRUCT(BlockDirective) {
+  CURRENT_FIELD(name, std::string);
+  CURRENT_FIELD(parameters, std::string);
+  CURRENT_FIELD(directives, std::vector<directives_t>);
+  CURRENT_DEFAULT_CONSTRUCTOR(BlockDirective) {}
+  CURRENT_CONSTRUCTOR(BlockDirective)(const std::string& name) : name(name) {}
+  CURRENT_CONSTRUCTOR(BlockDirective)(const std::string& name, const std::string& parameters)
+      : name(name), parameters(parameters) {}
+
+  SimpleDirective& AddSimpleDirective(const std::string& name, const std::string& parameters) {
+    directives.push_back(SimpleDirective(name, parameters));
+    return Value<SimpleDirective>(directives.back());
+  }
+
+  template <typename T>
+  T& AddDirective(T && directive) {
+    directives.push_back(std::forward<T>(directive));
+    return Value<T>(directives.back());
+  }
+};
+
+// clang-format off
+CURRENT_STRUCT(LocationDirective, BlockDirective) {
+  CURRENT_DEFAULT_CONSTRUCTOR(LocationDirective) : SUPER("location") {}
+  CURRENT_CONSTRUCTOR(LocationDirective)(const std::string& location) : SUPER("location", location) {}
+
+  LocationDirective& AddNestedLocation(const std::string& location) {
+    return SUPER::AddDirective(LocationDirective(location));
+  }
+};
+// clang-format on
+
+inline LocationDirective DefaultLocationDirective(const std::string& location) {
+  LocationDirective result(location);
+  result.AddSimpleDirective("proxy_set_header", "X-Real-IP $remote_addr");
+  result.AddSimpleDirective("proxy_set_header", "X-Forwarded-For $proxy_add_x_forwarded_for");
+  result.AddSimpleDirective("proxy_pass_request_headers", "on");
+  return result;
+}
+
+// clang-format off
+CURRENT_STRUCT(ServerDirective, BlockDirective) {
+  CURRENT_FIELD(port, uint16_t, 0u);
+  CURRENT_DEFAULT_CONSTRUCTOR(ServerDirective) : SUPER("server") {}
+  CURRENT_CONSTRUCTOR(ServerDirective)(uint16_t port) : SUPER("server"), port(port) {
+    SUPER::AddSimpleDirective("listen", current::ToString(port));
+  }
+
+  LocationDirective& AddLocation(const std::string& location) {
+    return SUPER::AddDirective(LocationDirective(location));
+  }
+
+  LocationDirective& AddDefaultLocation(const std::string& location) {
+    return SUPER::AddDirective(DefaultLocationDirective(location));
+  }
+};
+
+CURRENT_STRUCT(HttpDirective, BlockDirective) {
+  CURRENT_DEFAULT_CONSTRUCTOR(HttpDirective) : SUPER("http") {}
+
+  ServerDirective& AddServer(uint16_t port) {
+    for (const auto& d : SUPER::directives) {
+      if (Exists<ServerDirective>(d) && Value<ServerDirective>(d).port == port) {
+         CURRENT_THROW(PortAlreadyUsedException(port));
+      }
+    }
+    return SUPER::AddDirective(ServerDirective(port));
+  }
+};
+// clang-format on
+
+using full_config_directives_t = Variant<SimpleDirective, BlockDirective>;
+
+CURRENT_STRUCT(FullConfig) {
```

Comment:
Could `ServerConfig` or `ServiceConfig` or `HTTPServiceConfig` be a better name?